File: c++/medium/wordsearch.cpp
#include <iostream>

class Solution {
    bool search(vector<vector<char>> &board, string word, int row, int col, int index, int m, int n) {
        if(index == word.length()) {
            return true;
        } else if(row < 0 || col < 0 || row == m || col == n || board[row][col] != word[index] || board[row][col] == ' ') {
            return false;
        }

        char c = board[row][col];
        board[row][col] = ' ';

        bool up = search(board, word, row - 1, col, index + 1, m, n);
        bool right = search(board, word, row, col + 1, index + 1, m, n);
        bool down = search(board, word, row + 1, col, index + 1, m, n);
        bool left = search(board, word, row, col - 1, index + 1, m, n);

        board[row][col] = c;

        return up || right || down || left;
    }


public:
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size(), n = board[0].size();
        int index = 0;

        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(board[i][j] == word[index]) {
                    if(search(board, word, i, j, index, m, n)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }
};

File: python/medium/wordsearch.py
class Solution:
    def search(self, board: List[List[str]], word: str, row: int, col: int, index: int, m: int, n: int):
        if index == len(word) :
            return True
        elif(row < 0 or col < 0 or row == m or col == n or board[row][col] != word[index] or 
        board[row][col] == ''):  
            return False

        c = board[row][col]
        board[row][col] = ' '

        up = self.search(board, word, row - 1, col, index + 1, m, n)
        right = self.search(board, word, row, col + 1, index + 1, m, n)
        down = self.search(board, word, row + 1, col, index + 1, m, n)
        left = self.search(board, word, row, col - 1, index + 1, m, n)

        board[row][col] = c

        return up or right or down or left 

    def exist(self, board: List[List[str]], word: str):
        m = len(board)
        n = len(board[0])
        index = 0

        for i in range(m):
            for j in range(n):
                if board[i][j] == word[index]:
                    if self.search(board, word, i, j, index, m, n):
                        return True

        return False
        